File: rankboard-admin/server-python/app/services/excel_service.py
"""EXCEL BULK IMPORT — generate the sample template, and parse uploads.

Design choices worth understanding:

- We READ uploads in read_only mode and WRITE the sample with styling.
  openpyxl is the right tool here (not pandas) because we want precise
  control over headers, validation, and a styled template — and because
  it streams large files without loading everything into memory.

- Validation is PER ROW and never trusts the file. A user's spreadsheet
  is just untrusted input wearing a friendly extension: rows can be
  blank, mistyped, duplicated, or malicious. We collect the good rows
  and a per-row reason for every bad one, then report both — so the
  user fixes 3 bad rows instead of being told only "invalid file".

- The parser returns plain dicts; the ROUTE decides what to insert.
  Keeping DB writes out of here keeps this unit testable and focused.

- The user only ever supplies the KEYWORD. A keyword's Google position
  is discovered automatically the next time "Check rankings" runs — so
  the template has a single column and the parser ignores anything else.
"""
import io

from openpyxl import Workbook, load_workbook
from openpyxl.styles import Alignment, Font, PatternFill

# The contract between the sample file and the parser: this exact
# header. Keyword is the only thing we ask the user for.
COLUMNS = ["keyword"]
MAX_ROWS = 1000  # guard against someone uploading a 1M-row monster


def build_sample_workbook() -> bytes:
    """Return a styled .xlsx template as bytes, ready to stream to the
    browser. Row 1 = header, then a few example rows, then a notes
    block so the user knows the rules without reading docs."""
    wb = Workbook()
    ws = wb.active
    ws.title = "Keywords"

    header_fill = PatternFill("solid", start_color="EA580C")  # the app's orange
    header_font = Font(bold=True, color="FFFFFF", name="Arial")
    body_font = Font(name="Arial")

    # Header row
    ws.append(["Keyword"])
    cell = ws.cell(row=1, column=1)
    cell.fill = header_fill
    cell.font = header_font
    cell.alignment = Alignment(horizontal="left", vertical="center")
    ws.row_dimensions[1].height = 22

    # Example rows (clearly illustrative)
    examples = [
        "online yoga classes",
        "meditation retreat rishikesh",
        "pranayama breathing course",
    ]
    for i, kw in enumerate(examples):
        ws.append([kw])
        ws.cell(row=2 + i, column=1).font = body_font

    ws.column_dimensions["A"].width = 46

    # Notes a couple of rows below the data
    notes_start = 2 + len(examples) + 1
    notes = [
        "How to use this template:",
        "• Replace the example rows above with your own keywords — one per row.",
        "• Keyword: the search term you want to track (required).",
        "• You don't need to enter any rank — RankBoard finds each keyword's current position automatically the next time you run “Check rankings”.",
        "• Keep the header row. Delete these notes if you like.",
        "• Up to %d keywords per file." % MAX_ROWS,
    ]
    for i, line in enumerate(notes):
        cell = ws.cell(row=notes_start + i, column=1)
        cell.value = line
        cell.font = Font(name="Arial", italic=True, color="78716C", bold=(i == 0))

    buffer = io.BytesIO()
    wb.save(buffer)
    return buffer.getvalue()


def parse_keyword_workbook(file_bytes: bytes) -> tuple[list[dict], list[dict]]:
    """Parse an uploaded .xlsx. Returns (valid_rows, errors).

    valid_rows: [{"term"}, ...]   — rank columns are intentionally ignored
    errors:     [{"row": <excel row number>, "reason": "..."}, ...]
    """
    try:
        wb = load_workbook(io.BytesIO(file_bytes), read_only=True, data_only=True)
    except Exception:
        raise ValueError("That file couldn't be read as an Excel (.xlsx) workbook.")

    ws = wb.active
    valid: list[dict] = []
    errors: list[dict] = []
    seen_terms: set[str] = set()
    data_rows = 0

    # enumerate starting at row 1; we skip row 1 as the header
    for excel_row, row in enumerate(ws.iter_rows(values_only=True), start=1):
        if excel_row == 1:
            continue  # header
        if row is None:
            continue

        term_raw = row[0] if len(row) > 0 else None

        # Blank keyword → silently skip (trailing empties, notes, etc.).
        # Keyword is the only field, so a row with no keyword is just noise.
        if term_raw is None or str(term_raw).strip() == "":
            continue

        # Skip the template's own notes block if left in.
        if isinstance(term_raw, str) and term_raw.strip().startswith(("How to use", "•")):
            continue

        data_rows += 1
        if data_rows > MAX_ROWS:
            errors.append({"row": excel_row, "reason": f"File exceeds the {MAX_ROWS}-keyword limit; remaining rows ignored."})
            break

        term = str(term_raw).strip().lower()

        # Duplicate within the same file → keep the first, flag the rest.
        if term in seen_terms:
            errors.append({"row": excel_row, "reason": f"Duplicate of an earlier row for “{term}”; skipped."})
            continue
        seen_terms.add(term)

        valid.append({"term": term})

    return valid, errors

File: rankboard-admin/server-php-overlay/app/Services/ExcelService.php
<?php

namespace App\Services;

use InvalidArgumentException;
use PhpOffice\PhpSpreadsheet\IOFactory;
use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Style\Alignment;
use PhpOffice\PhpSpreadsheet\Style\Fill;
use Throwable;

/**
 * EXCEL BULK IMPORT — generate the sample template, and parse uploads.
 * A line-for-line port of the Python excel_service; same design:
 *
 * - Validation is PER ROW and never trusts the file. A spreadsheet is
 *   just untrusted input wearing a friendly extension. We collect the
 *   good rows and a per-row reason for every bad one, so the user
 *   fixes 3 rows instead of being told "invalid file".
 *
 * - The parser returns plain arrays; the CONTROLLER decides what to
 *   insert. Keeping DB writes out of here keeps this testable.
 */
class ExcelService
{
    /** The contract between the sample file and the parser. */
    public const COLUMNS = ['keyword', 'current_rank', 'previous_rank'];

    public const MAX_ROWS = 1000; // guard against a 1M-row monster

    public static function buildSampleSpreadsheet(): Spreadsheet
    {
        $spreadsheet = new Spreadsheet();
        $sheet = $spreadsheet->getActiveSheet();
        $sheet->setTitle('Keywords');

        // Header row
        $sheet->fromArray(['Keyword', 'Current Rank', 'Previous Rank (optional)'], null, 'A1');
        $sheet->getStyle('A1:C1')->applyFromArray([
            'font' => ['bold' => true, 'name' => 'Arial', 'color' => ['rgb' => 'FFFFFF']],
            'fill' => ['fillType' => Fill::FILL_SOLID, 'startColor' => ['rgb' => 'EA580C']], // the app's orange
            'alignment' => [
                'horizontal' => Alignment::HORIZONTAL_LEFT,
                'vertical' => Alignment::VERTICAL_CENTER,
            ],
        ]);
        $sheet->getRowDimension(1)->setRowHeight(22);

        // Example rows (clearly illustrative)
        $examples = [
            ['online yoga classes', 4, 9],
            ['meditation retreat rishikesh', 12, 8],
            ['pranayama breathing course', 3, null], // blank previous = "New"
        ];
        $sheet->fromArray($examples, null, 'A2');
        $sheet->getStyle('A2:C'.(1 + count($examples)))->applyFromArray([
            'font' => ['name' => 'Arial'],
        ]);

        $sheet->getColumnDimension('A')->setWidth(38);
        $sheet->getColumnDimension('B')->setWidth(16);
        $sheet->getColumnDimension('C')->setWidth(24);

        // Notes a couple of rows below the data
        $notesStart = 2 + count($examples) + 1;
        $notes = [
            'How to use this template:',
            '• Replace the example rows above with your own keywords.',
            '• Keyword: the search term (required).',
            '• Current Rank: a whole number, 1 or higher (required).',
            '• Previous Rank: a whole number, 1 or higher — leave blank for a brand-new keyword.',
            '• Keep the header row. Delete these notes if you like.',
            '• Up to '.self::MAX_ROWS.' keywords per file.',
        ];
        foreach ($notes as $i => $line) {
            $row = $notesStart + $i;
            $sheet->setCellValue('A'.$row, $line);
            $sheet->getStyle('A'.$row)->applyFromArray([
                'font' => [
                    'name' => 'Arial',
                    'italic' => true,
                    'bold' => $i === 0,
                    'color' => ['rgb' => '78716C'],
                ],
            ]);
        }

        return $spreadsheet;
    }

    /**
     * Parse an uploaded .xlsx from a file path.
     *
     * @return array{0: list<array{term: string, currentRank: int, previousRank: int|null}>,
     *               1: list<array{row: int, reason: string}>}  [valid, errors]
     */
    public static function parseKeywordWorkbook(string $path): array
    {
        try {
            $spreadsheet = IOFactory::load($path);
        } catch (Throwable $e) {
            throw new InvalidArgumentException("That file couldn't be read as an Excel (.xlsx) workbook.");
        }

        $sheet = $spreadsheet->getActiveSheet();
        // nullValue=null, calculate formulas, no display formatting,
        // numeric (0-based) column keys.
        $rows = $sheet->toArray(null, true, false, false);

        $valid = [];
        $errors = [];
        $seenTerms = [];
        $dataRows = 0;

        foreach ($rows as $idx => $row) {
            $excelRow = $idx + 1;
            if ($excelRow === 1) {
                continue; // header
            }

            $termRaw = $row[0] ?? null;
            $curRaw = $row[1] ?? null;
            $prevRaw = $row[2] ?? null;

            // Fully blank row -> silently skip (trailing empties, etc.)
            $termBlank = $termRaw === null || (is_string($termRaw) && trim($termRaw) === '');
            if ($termBlank && $curRaw === null && $prevRaw === null) {
                continue;
            }

            // Skip the template's own notes block if left in.
            if (is_string($termRaw)) {
                $trimmed = trim($termRaw);
                if (str_starts_with($trimmed, 'How to use') || str_starts_with($trimmed, '•')) {
                    continue;
                }
            }

            $dataRows++;
            if ($dataRows > self::MAX_ROWS) {
                $errors[] = ['row' => $excelRow, 'reason' => 'File exceeds the '.self::MAX_ROWS.'-keyword limit; remaining rows ignored.'];
                break;
            }

            $term = $termRaw === null ? '' : strtolower(trim((string) $termRaw));
            if ($term === '') {
                $errors[] = ['row' => $excelRow, 'reason' => 'Missing keyword.'];
                continue;
            }

            $cur = self::cleanInt($curRaw);
            if ($cur === null) {
                $errors[] = ['row' => $excelRow, 'reason' => 'Missing current rank.'];
                continue;
            }
            if ($cur === false || $cur < 1) {
                $errors[] = ['row' => $excelRow, 'reason' => 'Current rank must be a whole number of 1 or more (got '.self::repr($curRaw).').'];
                continue;
            }

            $prev = self::cleanInt($prevRaw);
            if ($prev === false || (is_int($prev) && $prev < 1)) {
                $errors[] = ['row' => $excelRow, 'reason' => 'Previous rank must be blank or a whole number of 1 or more (got '.self::repr($prevRaw).').'];
                continue;
            }
            // $prev is null (blank) or a valid int here.

            // Duplicate within the same file -> keep the first, flag the rest.
            if (isset($seenTerms[$term])) {
                $errors[] = ['row' => $excelRow, 'reason' => 'Duplicate of an earlier row for “'.$term.'”; skipped.'];
                continue;
            }
            $seenTerms[$term] = true;

            $valid[] = ['term' => $term, 'currentRank' => $cur, 'previousRank' => $prev];
        }

        return [$valid, $errors];
    }

    /**
     * Accept 5, 5.0, "5", " 5 " -> 5. Blank -> null. Junk -> false.
     * Mirrors the Python _clean_int, including its quirks.
     */
    private static function cleanInt(mixed $value): int|false|null
    {
        if ($value === null || (is_string($value) && trim($value) === '')) {
            return null; // blank
        }
        if (is_int($value)) {
            return $value;
        }
        if (is_float($value)) {
            // floats that aren't whole numbers (5.5) are invalid ranks
            return floor($value) == $value ? (int) $value : false;
        }
        if (is_string($value)) {
            $head = explode('.', trim($value))[0];
            if ($head !== '' && ctype_digit($head)) {
                return (int) $head;
            }
        }

        return false; // present but unparseable
    }

    /** Python-style repr for error messages ('abc' stays quoted, 5.5 doesn't). */
    private static function repr(mixed $v): string
    {
        if (is_string($v)) {
            return "'".$v."'";
        }
        if ($v === null) {
            return 'None';
        }
        if (is_bool($v)) {
            return $v ? 'True' : 'False';
        }

        return (string) $v;
    }
}
